lib/weka/core/instances.rb

```diff
@@ -226,6 +226,16 @@ def instance_from(instance_or_values, weight:)
           instance_or_values
         else
           data = internal_values_of(instance_or_values)
+          # string attribute has unlimited range of possible values.
+          # Check the return index, if it is -1 then add the value to
+          # the attribute before creating the instance
+          data.map!.with_index do |value, index|
+            if value == -1 and attribute(index).string?
+              attribute(index).add_string_value(instance_or_values[index].to_s)
+            else
+              value
+            end
+          end
```

Comment:
You are right–adding the string value should not be done in the Attribute’s `#internal_value_for` but already during the creation of the instance. Nice catch!